### ISSUE TITLE
load_folder() panics when given an absolute path but load() does not

### ISSUE THREAD

**Issue by iliags** [bevyengine/bevy#9458]
## Bevy version

0.11

The issue is currently present in the main branch as well: 
https://github.com/bevyengine/bevy/blob/b6a9d8eba7807b63be525e423911c0217706f13a/crates/bevy_asset/src/io/file_asset_io.rs#L112C4-L123C6

## What you did

```rust
server.load_folder("C:\\Absolute\\Path\\To\\Folder\\")
```

## What went wrong

```
thread 'Compute Task Pool (3)' panicked at 'called `Result::unwrap()` on an `Err` value: StripPrefixError(())', C:\<user>\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bevy_asset-0.11.0\src\io\file_asset_io.rs:120:47
```

- What were you expecting?
  - The folder contents are loaded into untyped handles or throw an error for invalid types
- What actually happened?
  - Bevy panics when given an absolute path to a folder

## Additional information
I'm using toml files as project files (like Unreal/Unity/Godot) and want to load and watch the files in the project folder for changes; this folder can be anywhere on the filesystem. The 'read_directory' function assumes that the provided path is relative to self and appends the provided path to the self root path which turns into "C:\\Path\\To\\Executable\\C:\\Absolute\\Path\\To\\Folder\\" and panics.

```rust
 fn read_directory(
        &self,
        path: &Path,
    ) -> Result<Box<dyn Iterator<Item = PathBuf>>, AssetIoError> {
        let root_path = self.root_path.to_owned();

        // This panics if path is absolute
        Ok(Box::new(fs::read_dir(root_path.join(path))?.map(
            move |entry| {
                let path = entry.unwrap().path();
                path.strip_prefix(&root_path).unwrap().to_owned()
            },
        )))
    }
```

The ```load``` function works as expected but the ```load_folder``` variation panics which is inconsistent behavior. I think the best fix would be to auto-detect absolute paths and add the root path if it's not. If that's not in the cards then an optional boolean parameter for ```is_absolute``` could work.

**Comment by alice-i-cecile** [comment]
What OS are you on? File system bugs are sometimes OS specific, so it's good to record it.

**Comment by iliags** [comment]
I'm on windows but it will have the same behavior on linux. The issue is that ```load_folder``` adds the root path regardless of if the provided path is absolute or not while ```load``` does not.

The workaround that I found is to set the root directory to the project path; if that's the intended use then maybe it would be better to add a line or two in the docs rather than a code fix.

**Comment by bushrat011899** [comment]
> he 'read_directory' function assumes that the provided path is relative to self and appends the provided path to the self root path which turns into "C:\Path\To\Executable\C:\Absolute\Path\To\Folder\" and panics.

I don't believe this is true. According to the [Path::join docs](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.join), `a.join(b)` will yield `b` if `b` is already an absolute path. The real error is that `Path::strip_prefix` will return an error if you attempt to strip a prefix that does not match the current path. You can't strip `a` from `b` if `b` does not already start with `a`.

To follow-up on your example:

```rust
let a: Path; // "C:\Path\To\Executable\"
let b: Path; // "C:\Absolute\Path\To\Folder\"

let c = a.join(b); // "C:\Absolute\Path\To\Folder\"
let d = c.strip_prefix(&a); // Error: `c` does not start with `a`
```

The minimal change that would resolve this bug would be to construct a relative path from `a` to `b` when `b.is_absolute()` is `true`. It looks like the most appropriate solution for that problem is using [`pathdiff::diff_paths`](https://docs.rs/pathdiff/latest/pathdiff/fn.diff_paths.html).

```rust
fn read_directory(
    &self,
    path: &Path,
) -> Result<Box<dyn Iterator<Item = PathBuf>>, AssetIoError> {
    let root_path = self.root_path.to_owned();

    // This panics if path is absolute
    Ok(Box::new(fs::read_dir(root_path.join(path))?.map(
        move |entry| {
            let path = entry.unwrap().path();
            pathdiff::diff_paths(path, root_path).unwrap()
        },
    )))
}
```

**Comment by iliags** [comment]
I'm new to rust so I'm not used to a sane(er) standard library so this is nice to learn.
